### ISSUE TITLE
Desiderata for terminal emulator capability reporting

### ISSUE THREAD

**Issue by gnachman** [irssi/irssi#1190]
This is perhaps an abuse of issues; please feel free to redirect me if I shouldn't be asking this here.

Some terminal emulator authors (VTE, xterm, tmux, mintty, libvterm, iTerm2, and others) are discussing building a new mechanism for reporting capabilities. For context: mintty/mintty#881

My goal is to collect desires from developers of popular applications. What capabilities do you wish you could discover about terminals that you don't already get from terminfo?

For example, being able to detect 24-bit color support, available cursor styles, bracketed paste support, and mouse reporting modes are the kinds of capabilities that would be included. They would likely be exposed through a combination of a new environment variable and a to-be-determined control sequence that reports them.

The results of this survey are being collected here: https://docs.google.com/spreadsheets/d/1H4OK8z98ptySnAPjlej56lrL6b0S1j6okwp7KLRTDBQ/edit#gid=0

Thanks in advance for your time!

**Comment by ailin-nemui** [comment]
hi, we already had TERMCAP environment variable and it failed, please don't repeat that. we also have TERM environment variable and it also failed (see browser user agents), so please don't repeat that either.

for example, you can check if xterm supports RGB using the `DCS + q (unpack "H*" "RGB") ST` sequence already. why not implement that everywhere?

Irssi could immediately be interested in querying, from your list, RGB, 256 colour, bracketed paste; for the future DECSLRM, notifications, focus reporting or clipboard access may be interesting. hyperlinks should also be added to your list

**Comment by dequis** [comment]
I agree with those points and would add: 

* immediately interested in: italics (because ever since i switched from TERM=screen to TERM=tmux remote servers that i ssh to stopped being aware of what my terminal is)
* future: sixel or whatever image protocol exists (i don't expect all terminals to agree on one protocol for this any time soon, but at least would be nice to query them)

**Comment by dequis** [comment]
BTW, thanks a lot to y'all for having these discussions. That mintty thread looks intense.

**Comment by gnachman** [comment]
> you can check if xterm supports RGB using the DCS + q (unpack "H*" "RGB") ST sequence already. why not implement that everywhere?

This is a better version of that DCS + q. It doesn't have a direct dependency on terminfo, which lets it evolve more quickly. It should also be faster because you won't have to query feature by feature—you can get all the feature flags in one request.

> italics (because ever since i switched from TERM=screen to TERM=tmux remote servers that i ssh to stopped being aware of what my terminal is)

Presumably the server lacks a terminfo entry for tmux. The problem with italics in particular is the xterm vs screen/tmux divide. Will discuss how to handle this.

> future: sixel or whatever image protocol exists (i don't expect all terminals to agree on one protocol for this any time soon, but at least would be nice to query them)

I believe there are quite a few at this point! Will investigate this too.

Thanks for the info.

**Comment by dequis** [comment]
> Presumably the server lacks a terminfo entry for tmux.

Yeah, which is why i think terminfo is not suitable for introducing new features at all, even simple ones like italics.

>The problem with italics in particular is the xterm vs screen/tmux divide.

Not quite, I get `\x1b[3m` for `sitm` in both `TERM=xterm` and `TERM=tmux` but nothing in `TERM=screen`. So more like a screen vs tmux divide, and tmux was introduced in terminfo 6.0.

That was 2015 and 5 years are still not enough for me to switch to a modern TERM because of the way it's propagated.
